File: Depth-First Search Algorithm/C++/Depth-First-Search-Algorithm.cpp
#include <iostream>
#include <vector>
#include <list>

class Graph {
public:
    Graph(int vertices) : vertices(vertices) {
        adj_list.resize(vertices);
        visited.resize(vertices, false);
    }

    void add_edge(int src, int dest) {
        adj_list[src].push_back(dest);
        adj_list[dest].push_back(src);
    }

    void dfs(int vertex) {
        visited[vertex] = true;
        std::cout << "Visited " << vertex << std::endl;

        for (int adj_vertex : adj_list[vertex]) {
            if (!visited[adj_vertex]) {
                dfs(adj_vertex);
            }
        }
    }

private:
    int vertices;
    std::vector<std::list<int>> adj_list;
    std::vector<bool> visited;
};

int main() {
    Graph g(6);

    g.add_edge(0, 1);
    g.add_edge(0, 2);
    g.add_edge(1, 3);
    g.add_edge(1, 4);
    g.add_edge(2, 4);
    g.add_edge(3, 5);
    g.add_edge(4, 5);

    g.dfs(0);

    return 0;
}

File: Depth-First Search Algorithm/Go/Depth-First-Search-Algorithm.go
package main

import (
	"fmt"
)

type Graph struct {
	vertices int
	adj_list []([]int)
	visited  []bool
}

func newGraph(vertices int) *Graph {
	return &Graph{
		vertices: vertices,
		adj_list: make([]([]int), vertices),
		visited:  make([]bool, vertices),
	}
}

func (g *Graph) addEdge(src, dest int) {
	g.adj_list[src] = append(g.adj_list[src], dest)
	g.adj_list[dest] = append(g.adj_list[dest], src)
}

func (g *Graph) dfs(vertex int) {
	g.visited[vertex] = true
	fmt.Printf("Visited %d\n", vertex)

	for _, adj_vertex := range g.adj_list[vertex] {
		if !g.visited[adj_vertex] {
			g.dfs(adj_vertex)
		}
	}
}

func main() {
	g := newGraph(6)

	g.addEdge(0, 1)
	g.addEdge(0, 2)
	g.addEdge(1, 3)
	g.addEdge(1, 4)
	g.addEdge(2, 4)
	g.addEdge(3, 5)
	g.addEdge(4, 5)

	g.dfs(0)
}
